File: Aug/Day30/valid_sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    if (!isValid(i, j, board, board[i][j])) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    private boolean isValid(int row, int col, char[][] board, char c) {
        for (int i = 0; i < 9; i++) {
            int l = 3 * (row / 3) + i / 3;
            int r = 3 * (col / 3) + i % 3;

            if (i != col && board[row][i] == c) return false;
            if (i != row && board[i][col] == c) return false;
            if (l != row && r != col && board[l][r] == c) return false;
        }
        return true;
    }
}

File: Aug/Day30/valid_sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        def isValid(row, col, board, c):
            for i in range(len(board)):
                l, r = (3 * (row//3)+ i//3) ,(3 * (col//3)+ i%3)
                if i!= col and board[row][i] == c:
                    return False
                elif i!= row and board[i][col] == c:
                    return False
                elif l!= row and r!= col and board[l][r] == c:
                    return False
            return True

        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j] != '.' :
                    if not isValid(i, j, board, board[i][j]):
                        return False
        return True